protocols/request-response/src/cbor.rs

```diff
@@ -0,0 +1,303 @@
+    ) -> io::Result<()>
+    where
+        T: AsyncWrite + Unpin + Send,
+    {
+        let data: Vec<u8> = serde_cbor::to_vec(&resp)
+            .map_err(|e| convert_error(e))
+            .unwrap();
+        write_length_prefixed(io, data.as_slice()).await?;
+        io.close().await?;
+
+        Ok(())
+    }
+}
+
+impl<Req, Resp> Behaviour<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned,
+    Resp: Send + Clone + Serialize + DeserializeOwned,
+{
+    pub fn new<I>(protocols: I, cfg: Config) -> Self
+    where
+        I: IntoIterator<
+            Item = (
+                <Codec<Req, Resp> as crate::Codec>::Protocol,
+                ProtocolSupport,
+            ),
+        >,
+    {
+        Behaviour {
+            inner: crate::Behaviour::new(
+                Codec {
+                    phantom: PhantomData,
+                },
+                protocols,
+                cfg,
+            ),
+        }
+    }
+
+    pub fn send_request(&mut self, peer: &PeerId, request: Req) -> RequestId {
+        self.inner.send_request(peer, request)
+    }
+
+    pub fn send_response(&mut self, ch: ResponseChannel<Resp>, rs: Resp) -> Result<(), Resp> {
+        self.inner.send_response(ch, rs)
+    }
+
+    pub fn add_address(&mut self, peer: &PeerId, address: Multiaddr) {
+        self.inner.add_address(peer, address)
+    }
+
+    pub fn remove_address(&mut self, peer: &PeerId, address: &Multiaddr) {
+        self.inner.remove_address(peer, address)
+    }
+
+    pub fn is_connected(&self, peer: &PeerId) -> bool {
+        self.inner.is_connected(peer)
+    }
+
+    pub fn is_pending_outbound(&self, peer: &PeerId, request_id: &RequestId) -> bool {
+        self.inner.is_pending_outbound(peer, request_id)
+    }
+
+    pub fn is_pending_inbound(&self, peer: &PeerId, request_id: &RequestId) -> bool {
+        self.inner.is_pending_inbound(peer, request_id)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::Codec;
+    use futures_ringbuf::Endpoint;
+    use libp2p_core::upgrade::{read_length_prefixed, write_length_prefixed};
+    use libp2p_swarm::StreamProtocol;
+    use serde::{Deserialize, Serialize};
+    use std::marker::PhantomData;
+
+    #[async_std::test]
+    async fn test_read_request() {
+        let (mut a, mut b) = Endpoint::pair(124, 124);
+        let expected = TestRequest {
+            payload: "test_payload".to_string(),
+        };
+        let data = serde_cbor::to_vec(&expected).expect("Should serialize");
+
+        write_length_prefixed(&mut a, data.as_slice())
+            .await
+            .expect("Should write");
+
+        let protocol = StreamProtocol::new("/test_cbor/1");
+        let mut codec: super::Codec<TestRequest, TestResponse> = super::Codec {
+            phantom: PhantomData,
+        };
+        let actual = codec
+            .read_request(&protocol, &mut b)
+            .await
+            .expect("Should read");
+
+        assert_eq!(actual, expected);
+    }
```

Comment:
Testing each function individually is a bit verbose. Can you merge them together and just test that a serialized request can be parsed again, essentially roundtripping the entire thing?

Like:

```rust
codec.write_request(a, req).await;
let actual = codec.read_request(b).await;

assert_eq(req, actual);

codec.write_response(b, res).await;
let actual = codec.read_response(b).await;

assert_eq(res, actual);
```